File: tools/dashboard/src/discovery.ts
import { readdirSync, existsSync, type Dirent } from "node:fs";
import { join, basename } from "node:path";
import { parseFrontmatter, parseStagesBlock, scanEntities } from "./parsing";
import type { Workflow, WorkflowData } from "./types";

export const IGNORED_DIRS = new Set([
  ".git", ".worktrees", "node_modules", "vendor", "dist", "build", "__pycache__", "tests",
]);

/**
 * Manual recursive walk that mimics Python os.walk() with dirnames pruning.
 * Does NOT use fs.readdirSync({recursive}) because that returns a flat list
 * with no ability to prune directories mid-walk.
 */
function walkDir(
  dir: string,
  callback: (dirPath: string, filenames: string[]) => void
): void {
  let entries: Dirent<string>[];
  try {
    entries = readdirSync(dir, { withFileTypes: true, encoding: "utf-8" }) as Dirent<string>[];
  } catch {
    return;
  }
  const subdirs: string[] = [];
  const filenames: string[] = [];
  for (const entry of entries) {
    if (entry.isDirectory()) {
      if (!IGNORED_DIRS.has(entry.name)) {
        subdirs.push(entry.name);
      }
    } else {
      filenames.push(entry.name);
    }
  }
  callback(dir, filenames);
  for (const sub of subdirs) {
    walkDir(join(dir, sub), callback);
  }
}

export function discoverWorkflows(root: string): Workflow[] {
  const workflows: Workflow[] = [];
  walkDir(root, (dirPath, filenames) => {
    if (!filenames.includes("README.md")) return;
    const readmePath = join(dirPath, "README.md");
    const fields = parseFrontmatter(readmePath);
    const commissionedBy = fields["commissioned-by"] ?? "";
    if (commissionedBy.startsWith("spacedock@")) {
      workflows.push({ dir: dirPath, commissioned_by: commissionedBy });
    }
  });
  return workflows;
}

export function aggregateWorkflow(workflowDir: string): WorkflowData | null {
  const readmePath = join(workflowDir, "README.md");
  if (!existsSync(readmePath)) return null;

  const fields = parseFrontmatter(readmePath);
  const stages = parseStagesBlock(readmePath) ?? [];
  const entities = scanEntities(workflowDir);

  const entityCountByStage: Record<string, number> = {};
  for (const e of entities) {
    const status = e.status;
    if (status) {
      entityCountByStage[status] = (entityCountByStage[status] ?? 0) + 1;
    }
  }

  return {
    dir: workflowDir,
    name: basename(workflowDir),
    commissioned_by: fields["commissioned-by"] ?? "",
    entity_type: fields["entity-type"] ?? "",
    entity_label: fields["entity-label"] ?? fields["entity-type"] ?? "entity",
    stages,
    entities,
    entity_count_by_stage: entityCountByStage,
  };
}

File: tools/dashboard/discovery.py
"""Workflow discovery -- finds Spacedock workflows by scanning for commissioned-by frontmatter.

Algorithm follows references/first-officer-shared-core.md:
- Recursively search for README.md files
- Check YAML frontmatter for 'commissioned-by: spacedock@...'
- Ignore .git, .worktrees, node_modules, vendor, dist, build, __pycache__
"""

import os

from tools.dashboard.parsing import parse_frontmatter, parse_stages_block, scan_entities

IGNORED_DIRS = {'.git', '.worktrees', 'node_modules', 'vendor', 'dist', 'build', '__pycache__', 'tests'}


def discover_workflows(root):
    """Recursively discover Spacedock workflow directories under root.

    Returns a list of dicts with 'dir' and 'commissioned_by' keys for each
    discovered workflow.
    """
    workflows = []
    for dirpath, dirnames, filenames in os.walk(root):
        dirnames[:] = [d for d in dirnames if d not in IGNORED_DIRS]

        if 'README.md' not in filenames:
            continue

        readme_path = os.path.join(dirpath, 'README.md')
        fields = parse_frontmatter(readme_path)
        commissioned_by = fields.get('commissioned-by', '')

        if commissioned_by.startswith('spacedock@'):
            workflows.append({
                'dir': dirpath,
                'commissioned_by': commissioned_by,
            })

    return workflows


def aggregate_workflow(workflow_dir):
    """Aggregate workflow data: stages, entities, and per-stage counts.

    Returns None if the directory has no README.md.
    """
    readme_path = os.path.join(workflow_dir, 'README.md')
    if not os.path.exists(readme_path):
        return None

    fields = parse_frontmatter(readme_path)
    stages = parse_stages_block(readme_path) or []
    entities = scan_entities(workflow_dir)

    entity_count_by_stage = {}
    for e in entities:
        status = e.get('status', '')
        if status:
            entity_count_by_stage[status] = entity_count_by_stage.get(status, 0) + 1

    return {
        'dir': workflow_dir,
        'name': os.path.basename(workflow_dir),
        'commissioned_by': fields.get('commissioned-by', ''),
        'entity_type': fields.get('entity-type', ''),
        'entity_label': fields.get('entity-label', fields.get('entity-type', 'entity')),
        'stages': stages,
        'entities': entities,
        'entity_count_by_stage': entity_count_by_stage,
    }
